File: Rudder-CustomerIO/Classes/RudderCustomerIOFactory.m
//
//  RudderCustomerIOFactory.m
//
//  Created by Arnab Pal on 29/10/19.
//

#import "RudderCustomerIOFactory.h"
#import "RudderCustomerIOIntegration.h"

@implementation RudderCustomerIOFactory

static RudderCustomerIOFactory *sharedInstance;

+ (instancetype)instance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (nonnull NSString *)key {
    return @"CustomerIO";
}

- (nonnull id<RSIntegration>)initiate:(nonnull NSDictionary *)config client:(nonnull RSClient *)client rudderConfig:(nonnull RSConfig *)rudderConfig {
    [RSLogger logDebug:@"Creating RudderIntegrationFactory"];
    return [[RudderCustomerIOIntegration alloc] initWithConfig:config withAnalytics:client withRudderConfig:rudderConfig];
}

@end

File: Rudder-CustomerIO/Classes/RudderCustomerIOFactory.swift
//
//  RudderCustomerIOFactory.swift
//  Rudder-CustomerIO
//
//  Created by Satheesh Kannan on 07/06/24.
//

import Foundation
import Rudder

// MARK: - RudderCustomerIOFactory
@objcMembers
public class RudderCustomerIOFactory: NSObject, RSIntegrationFactory {
    
    // MARK: - Instance
    public static let instance: RudderCustomerIOFactory = RudderCustomerIOFactory()
    
    // MARK: - Initializer
    public func initiate(_ config: [AnyHashable : Any], client: RSClient, rudderConfig: RSConfig) -> any RSIntegration {
        RSLogger.logDebug("Creating RudderIntegrationFactory")
        
        return RudderCustomerIOIntegration(config: config, client: client, rudderConfig: rudderConfig)
    }
    
    // MARK: - Key
    public func key() -> String {
        return "Customer IO"
    }
}
